// @did-btcr2/method/dist/cjs/core/beacon/aggregation/cohort/index.js
import { keyAggExport, keyAggregate, sortKeys } from '@scure/btc-signer/musig2';
import { payments, Transaction } from 'bitcoinjs-lib';
import { BeaconCoordinatorError } from '../../error.js';
import { BeaconCohortSigningSession } from '../session/index.js';
import { BeaconCohortReadyMessage } from './messages/keygen/cohort-ready.js';
import { COHORT_STATUS } from './status.js';
export class AggregateBeaconCohort {
    /**
     * Unique identifier for the cohort.
     * @type {string}
     */
    id;
    /**
     * DID of the coordinator.
     * @type {string}
     */
    coordinatorDid;
    /**
     * Minimum number of participants required to finalize the cohort.
     * @type {number}
     */
    minParticipants;
    /**
     * Status of the cohort.
     * @type {string}
     */
    status;
    /**
     * Network on which the cohort operates (e.g., 'mainnet', 'testnet').
     * @type {string}
     */
    network;
    /**
     * Pending signature requests, mapping participant DIDs to their pending signatures.
     * @type {Record<string, string>}
     */
    pendingSignatureRequests = {};
    /**
     * List of participant DIDs.
     * @type {Array<string>}
     */
    participants = new Array();
    /**
     * List of cohort keys.
     * @type {Array<Uint8Array>}
     */
    _cohortKeys = new Array();
    /**
     * Taproot Merkle root for the cohort.
     * @type {Uint8Array}
     */
    trMerkleRoot = new Uint8Array();
    /**
     * Beacon address for the cohort, calculated from the Taproot multisig.
     * @type {string}
     */
    beaconAddress = '';
    /**
     * Type of beacon used in the cohort (default is 'SMTBeacon').
     * @type {string}
     */
    beaconType;
    /**
     * Creates a new Musig2Cohort instance.
     * @param {Musig2CohortObject} params Parameters for initializing the cohort.
     * @param {string} [params.id] Optional unique identifier for the cohort. If not provided, a random UUID will be generated.
     * @param {number} params.minParticipants Minimum number of participants required to finalize the cohort.
     * @param {string} [params.coordinatorDid] DID of the coordinator managing the cohort.
     * @param {string} params.status Initial status of the cohort (e.g., 'PENDING', 'COHORT_SET').
     * @param {string} params.network Network on which the cohort operates (e.g., 'mainnet', 'testnet').
     */
    constructor({ id, minParticipants, coordinatorDid, status, network, beaconType }) {
        this.id = id || crypto.randomUUID();
        this.minParticipants = minParticipants || 2;
        this.coordinatorDid = coordinatorDid || '';
        this.status = status || COHORT_STATUS.COHORT_ADVERTISED;
        this.network = network;
        this.beaconType = beaconType || 'SMTBeacon';
    }
    set cohortKeys(keys) {
        this.cohortKeys = sortKeys(keys);
    }
    validateCohort(participantKeys, cohortKeys, beaconAddress) {
        for (const key of participantKeys) {
            if (!cohortKeys.includes(key)) {
                this.status = COHORT_STATUS.COHORT_FAILED;
                throw new BeaconCoordinatorError(`Participant key ${key} not found in cohort ${this.id}.`, 'COHORT_VALIDATION_ERROR', {
                    cohortId: this.id,
                    participantKeys: participantKeys,
                    cohortKeys: cohortKeys,
                    beaconAddress: beaconAddress
                });
            }
        }
        this.cohortKeys = cohortKeys.map(key => Buffer.from(key, 'hex'));
        const calculatedAddress = this.calulateBeaconAddress();
        if (calculatedAddress !== beaconAddress) {
            this.status = COHORT_STATUS.COHORT_FAILED;
            throw new BeaconCoordinatorError(`Calculated beacon address ${calculatedAddress} does not match provided address ${beaconAddress}.`, 'COHORT_ADDRESS_MISMATCH_ERROR', {
                cohortId: this.id,
                calculatedAddress,
                providedAddress: beaconAddress
            });
        }
        this.beaconAddress = beaconAddress;
        this.status = COHORT_STATUS.COHORT_SET_STATUS;
    }
    /**
     * Finalizes the cohort by checking if the minimum number of participants is met.
     * If the minimum is met, it sets the status to 'COHORT_SET_STATUS' and calculates the beacon address.
     * @throws {BeaconCoordinatorError} If the number of participants is less than the minimum required.
     * @returns {void}
     */
    finalize() {
        if (this.participants.length < this.minParticipants) {
            throw new BeaconCoordinatorError('Not enough participants to finalize the cohort', 'FINALIZE_COHORT_ERROR', {
                cohortId: this.id,
                participants: this.participants,
                minParticipants: this.minParticipants
            });
        }
        this.status = COHORT_STATUS.COHORT_SET_STATUS;
        this.beaconAddress = this.calulateBeaconAddress();
    }
    /**
     * Calculates the beacon Taproot multisig address for the cohort using participant keys.
     * @returns {string} The Taproot address for the cohort.
     * @throws {BeaconCoordinatorError} If the Taproot address cannot be calculated.
     */
    calulateBeaconAddress() {
        const keyAggContext = keyAggregate(this.cohortKeys);
        const aggPubkey = keyAggExport(keyAggContext);
        const payment = payments.p2tr({ internalPubkey: aggPubkey });
        if (!payment.hash) {
            throw new BeaconCoordinatorError('Failed to calculate Taproot Merkle root', 'CALCULATE_BEACON_MERKLE_ROOT_ERROR', {
                cohortId: this.id,
                cohortKeys: this.cohortKeys,
                minParticipants: this.minParticipants
            });
        }
        this.trMerkleRoot = payment.hash;
        if (!payment.address) {
            throw new BeaconCoordinatorError('Failed to calculate Taproot address', 'CALCULATE_BEACON_ADDRESS_ERROR', {
                cohortId: this.id,
                cohortKeys: this.cohortKeys,
                minParticipants: this.minParticipants
            });
        }
        return payment.address;
    }
    /**
     * Generates a CohortReadyMessage to be sent to participants when the cohort is set.
     * @param {string} to The DID of the participant to whom the message is sent.
     * @param {string} from The DID of the coordinator sending the message.
     * @returns {BeaconCohortReadyMessage} The CohortReadyMessage containing the cohort details.
     */
    getCohortReadyMessage(to, from) {
        if (this.status !== COHORT_STATUS.COHORT_SET_STATUS) {
            throw new BeaconCoordinatorError('Cohort status not "COHORT_SET".');
        }
        return new BeaconCohortReadyMessage({
            to,
            from,
            cohortId: this.id,
            beaconAddress: this.beaconAddress,
            cohortKeys: this.cohortKeys,
        });
    }
    /**
     * Adds a signature request to the pending requests for the cohort.
     * @param {BeaconCohortRequestSignatureMessage} message The signature request message to add.
     * @throws {Error} If a signature request from the same participant already exists.
     */
    addSignatureRequest(message) {
        if (!this.validateSignatureRequest(message)) {
            throw new BeaconCoordinatorError(`No signature request from ${message.from} in cohort ${this.id}.`);
        }
        if (!message.body?.data) {
            throw new BeaconCoordinatorError(`No signature data in request from ${message.from} in cohort ${this.id}.`);
        }
        this.pendingSignatureRequests[message.from] = message.body.data;
    }
    /**
     * Validates a signature request message to ensure it is from a participant in the cohort.
     * @param {BeaconCohortRequestSignatureMessage} message The signature request message to validate.
     * @returns {boolean} True if the message is valid, false otherwise.
     */
    validateSignatureRequest(message) {
        const cohortId = message.body?.cohortId;
        if (cohortId !== this.id) {
            console.info(`Signature request for wrong cohort: ${cohortId}.`);
            return false;
        }
        if (!this.participants.includes(message.from)) {
            console.info(`Participant ${message.from} not in cohort ${this.id}.`);
            return false;
        }
        return true;
    }
    /**
     * Starts a signing session for the cohort.
     * @returns {BeaconCohortSigningSession} The request signature message for the signing session.
     */
    startSigningSession() {
        console.debug(`Starting signing session for cohort ${this.id} with status ${this.status}`);
        if (this.status !== COHORT_STATUS.COHORT_SET_STATUS) {
            throw new BeaconCoordinatorError(`Cohort ${this.id} is not set.`);
        }
        // const smtRootBytes = new Uint8Array(32).map(() => Math.floor(Math.random() * 256));
        const cohort = new AggregateBeaconCohort(this);
        return new BeaconCohortSigningSession({
            cohort,
            pendingTx: new Transaction(),
            processedRequests: this.pendingSignatureRequests,
        });
    }
    /**
     * Converts the cohort instance to a JSON object representation.
     * @returns {BeaconCohort} The JSON object representation of the cohort.
     */
    json() {
        return Object.fromEntries(Object.entries(this));
    }
}
