// @did-btcr2/common/dist/cjs/logger.js
import chalk from 'chalk';
export var Env;
(function (Env) {
    Env["Development"] = "development";
    Env["Production"] = "production";
    Env["Test"] = "test";
})(Env || (Env = {}));
export const NODE_ENV = process.env.NODE_ENV || Env.Development;
/**
 * Mapping of log levels to colors for cleaner, more flexible logging.
 */
const LOG_LEVELS = {
    development: ['debug', 'info', 'log', 'warn', 'security'],
    test: ['log', 'info', 'error'],
    production: ['error'],
};
/**
 * Colors associated with each log level.
 */
const LEVEL_STYLES = {
    debug: chalk.green,
    error: chalk.red,
    info: chalk.blue,
    log: chalk.gray,
    warn: chalk.yellow,
    security: chalk.magenta,
};
/**
 * Defines the method mapping for console methods.
 */
// eslint-disable-next-line no-undef
const LEVEL_METHODS = {
    debug: 'debug',
    error: 'error',
    info: 'info',
    log: 'log',
    warn: 'warn',
    security: 'warn',
};
/**
 * A flexible, feature-rich logger with:
 * - Environment-based filtering
 * - Namespacing
 * - File/line tracing
 * - Timestamps
 * - Colorized output
 * @class Logger
 * @type {Logger}
 */
export class Logger {
    levels;
    namespace;
    useColors;
    static shared;
    /**
     * Creates a new Logger instance.
     * @param {string} namespace - Optional namespace for log messages.
     * @param {Object} options - Configuration options.
     * @param {Level[]} options.levels - Log levels to enable.
     * @param {boolean} options.useColors - Whether to use colored output.
     */
    constructor(namespace, options = {}) {
        this.levels = options.levels || LOG_LEVELS[NODE_ENV] || [];
        this.namespace = namespace || 'did-btcr2-js';
        const envForce = process.env.LOG_COLORS;
        this.useColors = options.useColors || (envForce ? envForce !== '0' && envForce.toLowerCase() !== 'false' : Boolean(process.stdout.isTTY));
    }
    /**
     * Logs a message with the specified level.
     * @param {Level} level - The log level.
     * @param {unknown} message - The message to log.
     * @param {...unknown[]} args - Additional arguments to log.
     * @returns {void}
     */
    _log(level, message, ...args) {
        if (!this.levels.includes(level))
            return;
        const color = LEVEL_STYLES[level];
        const method = LEVEL_METHODS[level];
        const timestamp = new Date().toISOString();
        const namespace = this.namespace ? `[${this.namespace}]` : '';
        const render = this.useColors ? color : (v) => v;
        const renderGray = this.useColors ? chalk.gray : (v) => v;
        console[method](`${renderGray(timestamp)} ${namespace} ${render(level)}: ${message}`, ...args);
    }
    debug(message, ...args) {
        this._log('debug', message, ...args);
        return this;
    }
    error(message, ...args) {
        this._log('error', message, ...args);
        return this;
    }
    info(message, ...args) {
        this._log('info', message, ...args);
        return this;
    }
    warn(message, ...args) {
        this._log('warn', message, ...args);
        return this;
    }
    security(message, ...args) {
        this._log('security', message, ...args);
        return this;
    }
    log(message, ...args) {
        this._log('log', message, ...args);
        return this;
    }
    newline() {
        console.log();
        return this;
    }
    /**
     * Static methods for convenience (auto-instantiate).
     * These use a shared singleton instance.
     * @param {unknown} message - The message to log.
     * @param {...unknown[]} args - Additional arguments to log.
     * @returns {void}
     */
    static debug(message, ...args) {
        Logger.instance().debug(message, ...args);
    }
    static error(message, ...args) {
        Logger.instance().error(message, ...args);
    }
    static info(message, ...args) {
        Logger.instance().info(message, ...args);
    }
    static warn(message, ...args) {
        Logger.instance().warn(message, ...args);
    }
    static security(message, ...args) {
        Logger.instance().security(message, ...args);
    }
    static log(message, ...args) {
        Logger.instance().log(message, ...args);
    }
    static newline() {
        Logger.instance().newline();
    }
    static instance(levels, useColors) {
        if (!Logger.shared) {
            Logger.shared = new Logger(undefined, { levels, useColors });
        }
        else {
            if (levels)
                Logger.shared.levels = levels;
            if (useColors !== undefined)
                Logger.shared.useColors = useColors;
        }
        return Logger.shared;
    }
}
