// @libp2p/kad-dht/dist/src/providers.js
import { PeerMap } from '@libp2p/peer-collections';
import * as varint from 'uint8-varint';
import { parseProviderKey, readProviderTime, toProviderKey } from './utils.js';
/**
 * Provides a mechanism to add and remove provider records from the datastore
 */
export class Providers {
    log;
    datastore;
    datastorePrefix;
    constructor(components, init) {
        this.log = components.logger.forComponent(`${init.logPrefix}:providers`);
        this.datastorePrefix = `${init.datastorePrefix}/provider`;
        this.datastore = components.datastore;
    }
    /**
     * Add a new provider for the given CID
     */
    async addProvider(cid, provider, options) {
        this.log.trace('%p provides %s', provider, cid);
        await this.writeProviderEntry(cid, provider, options);
    }
    /**
     * Remove a provider for the given CID
     */
    async removeProvider(cid, provider, options) {
        const key = toProviderKey(this.datastorePrefix, cid, provider);
        this.log.trace('%p no longer provides %s', provider, cid);
        await this.datastore.delete(key, options);
    }
    /**
     * Get a list of providers for the given CID
     */
    async getProviders(cid, options) {
        this.log.trace('get providers for %c', cid);
        const provs = await this.loadProviders(cid, options);
        this.log.trace('got %d providers for %c', provs.size, cid);
        return [...provs.keys()];
    }
    /**
     * Write a provider into the given store
     */
    async writeProviderEntry(cid, peerId, options) {
        const key = toProviderKey(this.datastorePrefix, cid, peerId);
        const buffer = varint.encode(options?.time?.getTime() ?? Date.now());
        await this.datastore.put(key, buffer, options);
    }
    /**
     * Load providers for the given CID from the store
     */
    async loadProviders(cid, options) {
        const providers = new PeerMap();
        const key = toProviderKey(this.datastorePrefix, cid);
        for await (const entry of this.datastore.query({ prefix: key.toString() }, options)) {
            const { peerId } = parseProviderKey(entry.key);
            providers.set(peerId, readProviderTime(entry.value));
        }
        return providers;
    }
}
