// @did-btcr2/method/dist/cjs/utils/did-document.js
import { getNetwork } from '@did-btcr2/bitcoin';
import { BTCR2_DID_DOCUMENT_CONTEXT, DidDocumentError, IdentifierTypes, INVALID_DID_DOCUMENT, JSONUtils } from '@did-btcr2/common';
import { CompressedSecp256k1PublicKey } from '@did-btcr2/keypair';
import { BeaconUtils } from '../core/beacon/utils.js';
import { Identifier } from '../core/identifier.js';
import { Appendix } from './appendix.js';
import { payments } from 'bitcoinjs-lib';
export const ID_PLACEHOLDER_VALUE = 'did:btcr2:_';
export const BECH32M_CHARS = '';
export const DID_REGEX = /did:btcr2:(x1[qpzry9x8gf2tvdw0s3jn54khce6mua7l]*)/g;
/**
 * DID BTCR2 Verification Method extends the DidVerificationMethod class adding helper methods and properties
 * @class DidVerificationMethod
 * @type {DidVerificationMethod}
 *
 */
export class DidVerificationMethod {
    id;
    type;
    controller;
    publicKeyMultibase;
    secretKeyMultibase;
    constructor({ id, type, controller, publicKeyMultibase, secretKeyMultibase }) {
        this.id = id;
        this.type = type;
        this.controller = controller;
        this.publicKeyMultibase = publicKeyMultibase;
        this.secretKeyMultibase = secretKeyMultibase;
        if (!secretKeyMultibase) {
            delete this.secretKeyMultibase;
        }
    }
}
/**
 * BTCR2 DID Document extends the DidDocument class adding helper methods and properties
 * @class DidDocument
 * @type {DidDocument}
 * @implements {Btcr2DidDocument}
 * @property {string} id - The identifier of the DID Document.
 * @property {Array<string>} [controller] - The controller of the DID Document.
 * @property {Array<string | JSONObject>} ['@context'] - The context of the DID Document.
 * @property {Array<DidVerificationMethod>} verificationMethod - The verification methods of the DID Document.
 * @property {Array<string | DidVerificationMethod>} [authentication] - The authentication methods of the DID Document.
 * @property {Array<string | DidVerificationMethod>} [assertionMethod] - The assertion methods of the DID Document.
 * @property {Array<string | DidVerificationMethod>} [capabilityInvocation] - The capability invocation methods of the DID Document.
 * @property {Array<string | DidVerificationMethod>} [capabilityDelegation] - The capability delegation methods of the DID Document.
 * @property {Array<BeaconService>} service - The services of the DID Document.
 */
export class DidDocument {
    id;
    controller;
    '@context' = BTCR2_DID_DOCUMENT_CONTEXT;
    verificationMethod;
    authentication;
    assertionMethod;
    capabilityInvocation;
    capabilityDelegation;
    service;
    deactivated;
    constructor(document) {
        // Set the ID and ID type
        const idType = document.id.includes('k')
            ? IdentifierTypes.KEY
            : IdentifierTypes.EXTERNAL;
        // Validate ID and parts for non-intermediate
        const isGenesis = document.id === ID_PLACEHOLDER_VALUE;
        // Deconstruct the document parts for validation
        const { id, controller, verificationMethod: vm, service } = document;
        // If not genesis, validate core properties
        if (!isGenesis) {
            if (!DidDocument.isValidId(id)) {
                throw new DidDocumentError(`Invalid id: ${id}`, INVALID_DID_DOCUMENT, document);
            }
            if (!DidDocument.isValidController(controller ?? [id])) {
                throw new DidDocumentError(`Invalid controller: ${controller}`, INVALID_DID_DOCUMENT, document);
            }
            if (!DidDocument.isValidVerificationMethods(vm)) {
                throw new DidDocumentError('Invalid verificationMethod: ' + vm, INVALID_DID_DOCUMENT, document);
            }
            if (!DidDocument.isValidServices(service)) {
                throw new DidDocumentError('Invalid service: ' + service, INVALID_DID_DOCUMENT, document);
            }
        }
        // Set core properties
        this.id = document.id;
        this.verificationMethod = document.verificationMethod;
        this.service = document.service;
        this['@context'] = document['@context'] || BTCR2_DID_DOCUMENT_CONTEXT;
        this.controller = document.controller || [this.id];
        // Relationships logic based on idType
        if (idType === IdentifierTypes.KEY) {
            // auto-generate #initialKey if missing
            const keyRef = `${this.id}#initialKey`;
            this.authentication = document.authentication || [keyRef];
            this.assertionMethod = document.assertionMethod || [keyRef];
            this.capabilityInvocation = document.capabilityInvocation || [keyRef];
            this.capabilityDelegation = document.capabilityDelegation || [keyRef];
        }
        else {
            // EXTERNAL: use provided arrays, must be defined
            this.authentication = document.authentication;
            this.assertionMethod = document.assertionMethod;
            this.capabilityInvocation = document.capabilityInvocation;
            this.capabilityDelegation = document.capabilityDelegation;
        }
        // Sanitize the DID Document
        DidDocument.sanitize(this);
        // If the DID Document is not an intermediateDocument, validate it
        if (!isGenesis) {
            DidDocument.validate(this);
        }
        else {
            this.validateGenesis();
        }
    }
    /**
     * Convert the DidDocument to a JSON object.
     * @returns {DidDocument} The JSON representation of the DidDocument.
     */
    json() {
        return Object.fromEntries(Object.entries(this));
    }
    /**
     * Create a minimal DidDocument from "k1" btcr2 identifier.
     * @param {string} publicKeyMultibase The public key in multibase format.
     * @param {Array<BeaconService>} service The beacon services to be included in the document.
     * @returns {DidDocument} A new DidDocument with the placeholder ID.
     */
    static fromKeyIdentifier(id, publicKeyMultibase, service) {
        // Ensure the ID is in the correct format
        id = id.includes('#') ? id : `${id}#initialKey`;
        // Create the verification method and the DidDocument
        const document = {
            id,
            verificationMethod: [
                new DidVerificationMethod({
                    id,
                    type: 'Multikey',
                    controller: id,
                    publicKeyMultibase
                })
            ],
            service
        };
        return new DidDocument(document);
    }
    /**
     * Create a DidDocument from "x1" btcr2 identifier.
     * @param {ExternalData} data The verification methods of the DID Document.
     * @returns {DidDocument} A new DidDocument.
     */
    static fromExternalIdentifier(data) {
        return new DidDocument(data);
    }
    /**
     * Sanitize the DID Document by removing undefined values
     * @returns {DidDocument} The sanitized DID Document
     */
    static sanitize(doc) {
        for (const key of Object.keys(doc)) {
            if (doc[key] === undefined) {
                delete doc[key];
            }
        }
        return doc;
    }
    /**
     * Validates a DidDocument by breaking it into modular validation methods.
     * @param {DidDocument} didDocument The DID document to validate.
     * @returns {boolean} True if the DID document is valid.
     * @throws {DidDocumentError} If any validation check fails.
     */
    static isValid(didDocument) {
        if (!this.isValidContext(didDocument?.['@context'])) {
            throw new DidDocumentError('Invalid "@context"', INVALID_DID_DOCUMENT, didDocument);
        }
        if (!this.isValidId(didDocument?.id)) {
            throw new DidDocumentError('Invalid "id"', INVALID_DID_DOCUMENT, didDocument);
        }
        if (!this.isValidVerificationMethods(didDocument?.verificationMethod)) {
            throw new DidDocumentError('Invalid "verificationMethod"', INVALID_DID_DOCUMENT, didDocument);
        }
        if (!this.isValidServices(didDocument?.service)) {
            throw new DidDocumentError('Invalid "service"', INVALID_DID_DOCUMENT, didDocument);
        }
        if (!this.isValidVerificationRelationships(didDocument)) {
            throw new DidDocumentError('Invalid verification relationships', INVALID_DID_DOCUMENT, didDocument);
        }
        return true;
    }
    /**
     * Validates that "@context" exists and includes correct values.
     * @private
     * @param {DidDocument['@context']} context The context to validate.
     * @returns {boolean} True if the context is valid.
     */
    static isValidContext(context) {
        if (!context)
            return false;
        if (!Array.isArray(context))
            return false;
        if (!context.every(ctx => typeof ctx === 'string' && BTCR2_DID_DOCUMENT_CONTEXT.includes(ctx)))
            return false;
        return true;
    }
    /**
     * Validates that the DID Document has a valid id.
     * @private
     * @param {string} id The id to validate.
     * @returns {boolean} True if the id is valid.
     */
    static isValidId(id) {
        if (typeof id !== 'string')
            return false;
        try {
            Identifier.decode(id);
            return true;
        }
        catch {
            return false;
        }
    }
    /**
     * Validates that the controller exists and is correctly formatted.
     * @param {Array<string>} controller The controller to validate.
     * @returns {boolean} True if the controller is valid.
     */
    static isValidController(controller) {
        if (!controller)
            return false;
        if (!Array.isArray(controller))
            return false;
        if (!controller.every(c => typeof c === 'string'))
            return false;
        return true;
    }
    /**
     * Validates that verification methods exist and are correctly formatted.
     * @private
     * @param {DidVerificationMethod[]} verificationMethod The verification methods to validate.
     * @returns {boolean} True if the verification methods are valid.
     */
    static isValidVerificationMethods(verificationMethod) {
        return Array.isArray(verificationMethod) && verificationMethod.every(Appendix.isDidVerificationMethod);
    }
    /**
     * Validates that the DID Document has valid services.
     * @private
     * @param {DidService[]} service The services to validate.
     * @returns {boolean} True if the services are valid.
     */
    static isValidServices(service) {
        return Array.isArray(service) && service.every(BeaconUtils.isBeaconService);
    }
    /**
     * Validates verification relationships (authentication, assertionMethod, capabilityInvocation, capabilityDelegation).
     * @private
     * @param {DidDocument} didDocument The DID Document to validate.
     * @returns {boolean} True if the verification relationships are valid.
     */
    static isValidVerificationRelationships(didDocument) {
        const possibleVerificationRelationships = [
            'authentication',
            'assertionMethod',
            'capabilityInvocation',
            'capabilityDelegation'
        ];
        const keys = Object.keys(didDocument);
        const availableKeys = possibleVerificationRelationships.filter(key => keys.includes(key));
        return availableKeys.every((key) => {
            const value = didDocument[key];
            return value &&
                Array.isArray(value) &&
                value.every((entry) => typeof entry === 'string' || Appendix.isDidVerificationMethod(entry));
        });
    }
    /**
     * Validate the DID Document
     * @returns {DidDocument} Validated DID Document.
     * @throws {DidDocumentError} If the DID Document is invalid.
     */
    static validate(didDocument) {
        // Validate the DID Document
        if (didDocument.id === ID_PLACEHOLDER_VALUE) {
            didDocument.validateGenesis();
        }
        else {
            DidDocument.isValid(didDocument);
        }
        // Return the DID Document
        return didDocument;
    }
    /**
     * Validate the GenesisDocument.
     * @returns {boolean} True if the GenesisDocument is valid.
     */
    validateGenesis() {
        // Validate the id
        if (this.id !== ID_PLACEHOLDER_VALUE) {
            throw new DidDocumentError('Invalid GenesisDocument ID', INVALID_DID_DOCUMENT, this);
        }
        // Validate the controller
        if (!this.controller?.every(c => c === ID_PLACEHOLDER_VALUE)) {
            throw new DidDocumentError('Invalid GenesisDocument controller', INVALID_DID_DOCUMENT, this);
        }
        // Validate the verificationMethod
        if (!this.verificationMethod.every(vm => vm.id.includes(ID_PLACEHOLDER_VALUE) && vm.controller === ID_PLACEHOLDER_VALUE)) {
            throw new DidDocumentError('Invalid GenesisDocument verificationMethod', INVALID_DID_DOCUMENT, this);
        }
        // Validate the service
        if (!this.service.every(svc => svc.id.includes(ID_PLACEHOLDER_VALUE))) {
            throw new DidDocumentError('Invalid GenesisDocument service', INVALID_DID_DOCUMENT, this);
        }
        if (!DidDocument.isValidVerificationRelationships(this)) {
            // Return true if all validations pass
            throw new DidDocumentError('Invalid GenesisDocument assertionMethod', INVALID_DID_DOCUMENT, this);
        }
        return true;
    }
    /**
     * Convert the DidDocument to an GenesisDocument.
     * @returns {GenesisDocument} The GenesisDocument representation of the DidDocument.
     */
    toIntermediate() {
        if (this.id.includes('k1')) {
            throw new DidDocumentError('Cannot convert a key identifier to an intermediate document', INVALID_DID_DOCUMENT, this);
        }
        return new GenesisDocument(this);
    }
}
export class Document {
    static isValid(didDocument) {
        return new DidDocument(didDocument).validateGenesis();
    }
}
/**
 * GenesisDocument extends the DidDocument class for creating and managing intermediate DID documents.
 * This class is used to create a minimal DID document with a placeholder ID.
 * It is used in the process of creating a new DID document.
 * @class GenesisDocument
 * @extends {DidDocument}
 */
export class GenesisDocument extends DidDocument {
    constructor(document) {
        super(document);
    }
    /**
     * Convert the GenesisDocument to a DidDocument by replacing the placeholder value with the provided DID.
     * @param did The DID to replace the placeholder value in the document.
     * @returns {DidDocument} A new DidDocument with the placeholder value replaced by the provided DID.
     */
    toDidDocument(did) {
        const stringThis = JSON.stringify(this).replaceAll(ID_PLACEHOLDER_VALUE, did);
        const parseThis = JSON.parse(stringThis);
        return new DidDocument(parseThis);
    }
    /**
     * Create an GenesisDocument from a DidDocument by replacing the DID with a placeholder value.
     * @param {DidDocument} didDocument The DidDocument to convert.
     * @returns {GenesisDocument} The GenesisDocument representation of the DidDocument.
     */
    static fromDidDocument(didDocument) {
        const intermediateDocument = JSONUtils.cloneReplace(didDocument, DID_REGEX, ID_PLACEHOLDER_VALUE);
        return new GenesisDocument(intermediateDocument);
    }
    /**
     * Create a minimal GenesisDocument with a placeholder ID.
     * @param {Array<DidVerificationMethod>} verificationMethod The public key in multibase format.
     * @param {VerificationRelationships} relationships The public key in multibase format.
     * @param {Array<BeaconService>} service The service to be included in the document.
     * @returns {GenesisDocument} A new GenesisDocument with the placeholder ID.
     */
    static create(verificationMethod, relationships, service) {
        const id = ID_PLACEHOLDER_VALUE;
        return new GenesisDocument({ id, ...relationships, verificationMethod, service, });
    }
    /**
     * Create a minimal GenesisDocument from a public key.
     * @param {KeyBytes} publicKey The public key in bytes format.
     * @returns {GenesisDocument} A new GenesisDocument with the placeholder ID.
     */
    static fromPublicKey(publicKey, network) {
        const pk = new CompressedSecp256k1PublicKey(publicKey);
        const id = ID_PLACEHOLDER_VALUE;
        const address = payments.p2pkh({ pubkey: pk.compressed, network: getNetwork(network) })?.address;
        const services = [{
                id: `${id}#key-0`,
                serviceEndpoint: `bitcoin:${address}`,
                type: 'SingletonBeacon',
            }];
        const relationships = {
            authentication: [`${id}#key-0`],
            assertionMethod: [`${id}#key-0`],
            capabilityInvocation: [`${id}#key-0`],
            capabilityDelegation: [`${id}#key-0`]
        };
        const verificationMethod = [
            {
                id: `${id}#key-0`,
                type: 'Multikey',
                controller: id,
                publicKeyMultibase: pk.multibase.encoded,
            }
        ];
        return GenesisDocument.create(verificationMethod, relationships, services);
    }
    /**
     * Taken an object, convert it to an IntermediateDocuemnt and then to a DidDocument.
     * @param {object | DidDocument} object The JSON object to convert.
     * @returns {DidDocument} The created DidDocument.
     */
    static fromJSON(object) {
        return new GenesisDocument(object);
    }
}
