// @did-btcr2/bitcoin/dist/esm/types.js
;
export class RestClientConfig {
    host;
    headers;
    constructor({ host, headers }) {
        this.host = host;
        this.headers = headers;
    }
}
;
export class BitcoinRpcClientConfig {
    network;
    headers;
    host;
    logger;
    password;
    timeout;
    username;
    version;
    wallet;
    allowDefaultWallet;
    constructor(options = {
        headers: {},
        host: 'http://localhost',
        logger: console,
        password: '',
        timeout: 30000,
        username: '',
        version: '0.21.1',
        wallet: '',
        allowDefaultWallet: false,
    }) {
        this.headers = options.headers;
        this.host = options.host;
        this.logger = options.logger;
        this.password = options.password;
        this.timeout = options.timeout;
        this.username = options.username;
        this.version = options.version;
        this.wallet = options.wallet;
        this.allowDefaultWallet = options.allowDefaultWallet;
    }
}
;
;
/**
 * Defines verbosity levels for block and transaction outputs.
 * Used to specify the format of returned block or transaction data.
 * @enum {number} VerbosityLevel for block and transaction outputs.
 */
export var VerbosityLevel;
(function (VerbosityLevel) {
    /** Return block or transaction data in raw hex-encoded format */
    VerbosityLevel[VerbosityLevel["hex"] = 0] = "hex";
    /** Return block or transaction data in JSON object format */
    VerbosityLevel[VerbosityLevel["json"] = 1] = "json";
    /**
     * Return block or transaction data in JSON object format with additional information.
     * Returns block data with information about each transaction.
     * Returns transaction data with information about the transaction including fee and prevout information.
     */
    VerbosityLevel[VerbosityLevel["jsonext"] = 2] = "jsonext";
    /**
     * Return block data in JSON object format with additional information.
     * Returns block data with information about each transaction, including prevout information for inputs.
     */
    VerbosityLevel[VerbosityLevel["jsonextprev"] = 3] = "jsonextprev";
})(VerbosityLevel || (VerbosityLevel = {}));
