// @did-btcr2/method/dist/cjs/core/beacon/error.js
import { MethodError } from '@did-btcr2/common';
export class BeaconError extends MethodError {
    constructor(message, type = 'BeaconError', data) {
        super(message, type, data);
    }
}
export class BeaconCoordinatorError extends MethodError {
    constructor(message, type = 'BeaconCoordinatorError', data) {
        super(message, type, data);
    }
}
export class BeaconParticipantError extends MethodError {
    constructor(message, type = 'BeaconParticipantError', data) {
        super(message, type, data);
    }
}
export class SingletonBeaconError extends MethodError {
    constructor(message, type = 'SingletonBeaconError', data) {
        super(message, type, data);
    }
}
export class AggregateBeaconError extends MethodError {
    constructor(message, type = 'AggregateBeaconError', data) {
        super(message, type, data);
    }
}
export class CASBeaconError extends MethodError {
    constructor(message, type = 'CASBeaconError', data) {
        super(message, type, data);
    }
}
export class SMTBeaconError extends MethodError {
    constructor(message, type = 'SMTBeaconError', data) {
        super(message, type, data);
    }
}
