// @did-btcr2/kms/dist/esm/store.js
/**
 * Re-implementation of a simple in-memory key-value store.
 *
 * The `MemoryStore` class is an implementation of
 * `KeyValueStore` that holds data in memory.
 *
 * It provides a basic key-value store that works synchronously and keeps all
 * data in memory. This can be used for testing, or for handling small amounts
 * of data with simple key-value semantics.
 *
 * Example usage:
 *
 * ```ts
 * const memoryStore = new MemoryStore<string, number>();
 * await memoryStore.set("key1", 1);
 * const value = await memoryStore.get("key1");
 * console.log(value); // 1
 * ```
 *
 * @public
 */
export class MemoryStore {
    /**
     * A private field that contains the Map used as the key-value store.
     */
    store = new Map();
    /**
     * Clears all entries in the key-value store.
     *
     * @returns {void} returns once the operation is complete.
     */
    clear() {
        this.store.clear();
    }
    /**
     * This operation is no-op for `MemoryStore`.
     */
    close() {
        /** no-op */
    }
    /**
     * Deletes an entry from the key-value store by its key.
     *
     * @param {K} id - The key of the entry to delete.
     * @returns {boolean} a boolean indicating whether the entry was successfully deleted.
     */
    delete(id) {
        return this.store.delete(id);
    }
    /**
     * Retrieves the value of an entry by its key.
     *
     * @param {K} id - The key of the entry to retrieve.
     * @returns {V | undefined} the value of the entry, or `undefined` if the entry does not exist.
     */
    get(id) {
        return this.store.get(id);
    }
    /**
     * Checks for the presence of an entry by key.
     *
     * @param {K} id - The key to check for the existence of.
     * @returns {boolean} a boolean indicating whether an element with the specified key exists or not.
     */
    has(id) {
        return this.store.has(id);
    }
    /**
     * Retrieves all values in the key-value store.
  
     * @returns {Array<V>} an array of all values in the store.
     */
    list() {
        return Array.from(this.store.values());
    }
    /**
     * Retrieves all entries in the key-value store.
     *
     * @returns {Array<[K, V]>} an array of key-value pairs in the store.
     */
    entries() {
        return Array.from(this.store.entries());
    }
    /**
     * Sets the value of an entry in the key-value store.
     *
     * @param {K} id - The key of the entry to set.
     * @param {V} key - The new value for the entry.
     * @returns {void} once operation is complete.
     */
    set(id, key) {
        this.store.set(id, key);
    }
}
