// @did-btcr2/method/dist/esm/utils/did-document-builder.js
import { DidDocumentError, INVALID_DID_DOCUMENT } from '@did-btcr2/common';
import { DidDocument } from './did-document.js';
/**
 * A builder class for constructing DID Documents.
 * @type {DidDocumentBuilder}
 * @class DidDocumentBuilder
 */
export class DidDocumentBuilder {
    document = {};
    constructor(initialDocument) {
        if (!initialDocument.id) {
            throw new DidDocumentError('Missing required "id" property', INVALID_DID_DOCUMENT, initialDocument);
        }
        this.document.id = initialDocument.id;
        this.document.verificationMethod = initialDocument.verificationMethod ?? [];
        if (initialDocument['@context']) {
            this.document['@context'] = initialDocument['@context'];
        }
    }
    withController(controller) {
        if (controller) {
            this.document.controller = controller ?? [this.document.id];
        }
        return this;
    }
    withAuthentication(authentication) {
        if (authentication) {
            this.document.authentication = authentication;
        }
        return this;
    }
    withAssertionMethod(assertionMethod) {
        if (assertionMethod) {
            this.document.assertionMethod = assertionMethod;
        }
        return this;
    }
    withCapabilityInvocation(capabilityInvocation) {
        if (capabilityInvocation) {
            this.document.capabilityInvocation = capabilityInvocation;
        }
        return this;
    }
    withCapabilityDelegation(capabilityDelegation) {
        if (capabilityDelegation) {
            this.document.capabilityDelegation = capabilityDelegation;
        }
        return this;
    }
    withService(service) {
        if (service) {
            this.document.service = service;
        }
        return this;
    }
    build() {
        const didDocument = new DidDocument(this.document);
        for (const key of Object.keys(didDocument)) {
            if (didDocument[key] === undefined) {
                delete didDocument[key];
            }
        }
        return didDocument;
    }
}
